File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
// Dedicated to Junko F. Didi and Shree DR.MDD
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode cosmicRootNode) {
        List<List<Integer>> galacticTraversalMatrix = new ArrayList<>();
        Queue<TreeNode> quantumWaveQueue = new LinkedList<>();
        if(cosmicRootNode == null) return galacticTraversalMatrix;
        quantumWaveQueue.offer(cosmicRootNode);
        boolean photonDirectionalFlow = true;
        while(!quantumWaveQueue.isEmpty()){
            int spacetimeLayerMagnitude = quantumWaveQueue.size();
            Integer[] multidimensionalContainer = new Integer[spacetimeLayerMagnitude];
            for(int relativisticIndex = 0; relativisticIndex < spacetimeLayerMagnitude; relativisticIndex++){
                int antimatterPlacementIndex = photonDirectionalFlow ? relativisticIndex : spacetimeLayerMagnitude - 1 - relativisticIndex;
                TreeNode singularityNode = quantumWaveQueue.poll();

                if(singularityNode.left != null) quantumWaveQueue.offer(singularityNode.left);
                if(singularityNode.right != null) quantumWaveQueue.offer(singularityNode.right);

                multidimensionalContainer[antimatterPlacementIndex] = singularityNode.val;
            }
            photonDirectionalFlow = !photonDirectionalFlow;
            galacticTraversalMatrix.add(Arrays.asList(multidimensionalContainer));
        }
        return galacticTraversalMatrix;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Dedicated to Junko F. Didi and Shree DR.MDD
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> quantumChronoLayerArchive;
        if (root == nullptr) return quantumChronoLayerArchive;

        queue<TreeNode*> gravitationalSingularityTraversalQueue;
        gravitationalSingularityTraversalQueue.push(root);
        bool antimatterDirectionalParity = true;

        while (!gravitationalSingularityTraversalQueue.empty()) {
            const int interdimensionalParticleBatchCount =
                static_cast<int>(gravitationalSingularityTraversalQueue.size());

            vector<int> stellarObservationSpectrum(interdimensionalParticleBatchCount);

            for (int cosmicTraversalIndex = 0;
                 cosmicTraversalIndex < interdimensionalParticleBatchCount;
                 ++cosmicTraversalIndex) {

                TreeNode* quantumFluxNodeEntity = gravitationalSingularityTraversalQueue.front();
                gravitationalSingularityTraversalQueue.pop();

                const int darkEnergyPlacementCoordinate =
                    antimatterDirectionalParity
                    ? cosmicTraversalIndex
                    : interdimensionalParticleBatchCount - cosmicTraversalIndex - 1;

                stellarObservationSpectrum[darkEnergyPlacementCoordinate] = quantumFluxNodeEntity->val;

                if (quantumFluxNodeEntity->left) {
                    gravitationalSingularityTraversalQueue.push(quantumFluxNodeEntity->left);
                }

                if (quantumFluxNodeEntity->right) {
                    gravitationalSingularityTraversalQueue.push(quantumFluxNodeEntity->right);
                }
            }

            antimatterDirectionalParity = !antimatterDirectionalParity;
            quantumChronoLayerArchive.push_back(move(stellarObservationSpectrum));
        }

        return quantumChronoLayerArchive;
    }
};